src/librustc_mir/interpret/operand.rs

```diff
@@ -593,7 +593,13 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
             let ty = self.monomorphize(val.ty)?;
             self.layout_of(ty)
         })?;
-        let op = match val.val {
+        let val = match val.val {
+            ConstValue::Param(_) => self.monomorphize(val)?.val,
```

Comment:
At this point you can just replace the entire match with `self.monomorphize(val)?.val`. Since the above also monomorphizes `val.ty`, you could call `let val = self.monomorphize(val)?;` before that and thus only have a single invocation of the substitution code.